[CM] Allow Central management to send errors or events back to the Kibana backend

ph opened the pull request:
# NOTES: DO NOT MERGE IT DEPENDS ON FEATURE OF KIBANA

Motivation: Currently CM  send on the next tick a ConfigOK => true if the previous config was successfully applied, the problem is this information is just 0 or 1 and in the context of beat we need a bit more details to make correct decision.


## Changes:

This PR do the following thing:

1. Add a AuthClient that wraps a Client but make sure that all calls are authenticated using the defined UUID and the AccessToken.

2. Add an EventsReporter as part of the `api` package that will send asynchrously api.Event to the CM backend. The reporter is started and closed at the same time of the Manager, on close events are flushed to the backend.

3. Events are send in batch of a 1000 to the Kibana endpoint. (configurable)

5. Any errors from `BlackList` or `Reload` are send as Event.

6. Introduce Manager's State with the following: STARTING, IN_PROGRESS, RUNNING, FAILED, STOPPED and will be send to the event endpoints.

7. Events are send by default every 30 sec. (configurable)



## Requests:

PATH: /api/beats/{UUID}/events exist
Verb used: "POST"
*uuid could be omnitted*

Request example:
```json
[
{"type": "STATE", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "STARTING", "message": "State change: STARTING"}},
{"type": "STATE", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "In_PROGRESS", "message": "State change: IN_PROGRESS"}},
{"type": "ERROR", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "CONFIG", "message": "a bad error", "uuid": "xxxxxxxxxxxxx"}} 
{"type": "ERROR", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "CONFIG", "message": "another bad error", "uuid": "xxxxxxxxxxxxx"}}
{"type": "STATE", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "FAILED", "message": "State change: FAILED"}},
{"type": "STATE", "timestamp": "2019-01-04T14:32:03.36764-05:00", "event": { "type": "STOPPED", "message": "State change: STOPPED"}},
]
```

Expected response:
Status code : 200
```json
{
"response": [
  { "success": true }
  { "success": true }
  { "success": true }
  { "success": true }
  { "success": true }
  { "success": true }
]
}
```
## TODO:
- [ ] Add the UUID from the API, waiting on api implementation to do that / review.
- [ ] Add Severity to the `ERROR`, "Fatal" or "Transient"


## Questions
- [ ] Should we create a UUID for each events and send them as part of the response?
- [ ] The above could allow us to ACK a batch of events.

Fixes: #9382

ph commented:
Add level to the config, right they are all fatal and not transient.

urso commented on x-pack/libbeat/management/api/client.go:
```diff
@@ -60,18 +59,23 @@ func NewClient(cfg *kibana.ClientConfig) (*Client, error) {
 // do a request to the API and unmarshall the message, error if anything fails
 func (c *Client) request(method, extraPath string,
-	params common.MapStr, headers http.Header, message interface{}) (int, error) {
+	body interface{}, headers http.Header, message interface{}) (int, error) {
```
nit: i'd prefer 'headers' before 'body' :)

What is the difference between 'body' and 'message' ?

urso replied:
Oh, I see. `message` is the response structure. Maybe we should call it resp then? It should also be clear the it must be a pointer.

ph replied:
I can change the order and naming, +1 :)

ph replied:
I will do the ordering in another PR, that will break my other in flight PR :(

ph commented on x-pack/libbeat/management/error.go:
```diff
@@ -0,0 +1,108 @@
+	Type ErrorType
+	UUID uuid.UUID
+	Err  error
+}
```
I think I will change that and make `Err` just a string instead it will be more appropriate.

urso replied:
If you turn it into a string, maybe rename the field to 'Message' ?

urso commented on x-pack/libbeat/management/api/event_reporter.go:
```diff
@@ -0,0 +1,131 @@
+			e.reportEvents(buffer)
+			buffer = nil
+		}
+	}
```
The buffer is unbounded, which means we have no control over memory usage. If one event producer becomes too chatty we either have to drop or block.

The channel itself is unbuffered. The effect of the channel being unbuffered will be back-pressure/blocking whenever we publish.

Alternative implementation without blocking:

```
type EventReporter struct {
  done chan struct{}

  buffer []Event
  mu sync.Mutex
}

func (e *EventReporter) AddEvent(evt Event) {
  e.mu.Lock()
  e.buffer = append(e.buffer, evt)
  e.mu.Unlock()
}

func (e *EventReporter) worker() {
  var done bool

  for !done {
    select {
      case <-e.done:
        done = true
      case <-ticker.C:
    }

    // flush buffer:
    var buf []Event
    e.mu.Lock()
    buf, e.buffer = e.buffer, nil
    e.mu.Unlock()
    e.reportEvents(buf)
  }
}
```

Instead of clearing the buffer one can use double-buffering and set `e.buffer = e.buffers[e.last][:0]`, so to keep the capacity.

ph replied:
@urso I was wondering if we should instead use a buffered channel set with the limit of the batch size, which should keep memory under some limits, at most we would have 2 batches in memory.

ph replied:
Another thing we would do could be to actually send events either on the number of events or the ticker trigger.

ph replied:
The problem with the above is chatty clients could be a problem, in that case we might consider throttling somehow the events.

urso replied:
Is there a notion of event priority? Maybe we could filter/remove some events if we need to make space.
I'm also afraid of potentially blocking the beat if we integrate this with logging.

ph replied:
@urso Having priority sound good, lets do it when we integrate with logging.

urso commented on x-pack/libbeat/management/api/event_reporter.go:
```diff
@@ -0,0 +1,131 @@
+		if err := e.sendEvents(events[pos:j]); err != nil {
+			errors = append(errors, err)
+		}
+	}
```
alternative implementations splitting the slice + using objects instead of indices:

```
for len(events) > e.maxBatchSize {
  batch := events[:e.maxBatchSize]
  events = events[e.maxBatchSize:]
  if err := e.sendEvents(batch); err != nil {
    errors = append(errors, err)
  }
}
if len(events) > 0 {
  if err := e.sendEvents(events); err != nil {
    errors = append(errors, err)
  }
}
```

or 

```
for len(events) > 0 {
  batch := events
  if len(batch) > e.maxBatchSize {
    batch = batch[:e.maxBatchSize]
  }

  if err := e.sendEvents(batch); err != nil {
    errors = append(errors, err)
  }

  events = events[len(batch):]
}
```

P.S: I don't mind using indices.

ph commented:
@urso I've made the required changes an made the event reporter async, much better :)

ph commented:
I have fixed the test and rebased

mattapperson commented:
@ph a few additions to the API:
- If the beat ID provided in the URL is invalid, the server will return a 400 error.
- If token auth fails, server will return  401
- If there is an issue with an event, the server will respond with:
```js
{
  response: [
    {success: false: reason: string}
  ]
}

ph commented:
Good I will make the change for response the 400 and 401 should be handled by the generic error handling but I will take a look.

ph commented:
I am ok to get that merge in the current state.

I have adjusted the PR to some fields coming from kibana and also fixed used the `_subtype` field in this PR.
